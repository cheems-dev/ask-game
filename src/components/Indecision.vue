<template>
  <img v-if="image" :src="image" alt="bg">
  <div class="bg-dark"></div>

  <div class="indecision-container">
    <h1>Recuerda terminar con un signo de interrogaci√≥n (?)</h1>
    <input type="text" placeholder="Haz una pregunta" v-model="question ">
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h2>{{ answer }}</h2>
      <h2 v-if="!answer" class="center"> 
        <HollowDotsSpinner/>
      </h2>
    </div>
  </div>
</template>

<script>
import HollowDotsSpinner from './HollowDotsSpinner.vue'

export default {
  name: 'Indecision',
  data(){
    return {
      question : null,
      answer: null,
      image: null,
      isValidQuestion: false,
    }
  },
  methods: {
    async getAnswer() {
      try {
        const {answer, image} = await fetch('https://yesno.wtf/api').then(
          response => response.json()
        )
        this.isValidQuestion = true
        this.answer = (answer === 'yes') ? 'Si' : 'No'
        this.image = image
      } catch (error) {
        console.log(`IndecisionComponent: ${error}`)        
        this.answer = 'No se pudo cargar la API'
        this.image = null
      }
    }
  },
  watch: {
    question(value, oldValue){
      this.isValidQuestion = false
      console.log({value})
      if(!value.includes('?'))    
        return
      this.isValidQuestion = true
      this.getAnswer()
    }
  },
  components: {
    HollowDotsSpinner,
  }
}
</script>

<style scoped>
    
    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.85);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        -webkit-appearance: none;
        -moz-appearance: none;
        appearance: none;
        font: 1.5vw/1  'Special Elite', cursive;
        font-size: 1.5vmin 3vh 4vmax;
        color: #333;
        -webkit-font-smoothing: antialiased;
        -moz-osx-font-smoothing: grayscale;
        width: 100%;
        max-width: 60vw;
        background-color: #ddd;
        border: none;
        padding: 10px 11px 11px 11px;
        border-radius: 3px;
        box-shadow: none;
        outline: none;
        margin: 0;
        box-sizing: border-box;
        border-style: dashed;
        text-align: center;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }

    h1{
      font-size: 3vw;
    }
    
    h2 {
      margin-top: 10vh;
      font-size: 3vw;
    }

    .center {
      display: flex;
      justify-content: center;
    }

  
</style>  