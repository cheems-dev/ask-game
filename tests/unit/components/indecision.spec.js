import Indecision from '@/components/Indecision.vue'
import { shallowMount } from '@vue/test-utils'

describe('Indecision Component', () => {
  let wrapper
  let clgSpy

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: 'yes',
          forced: false,
          image: 'https://yesno.wtf/assets/yes/2.gif',
        }),
    })
  )

  beforeEach(() => {
    wrapper = shallowMount(Indecision)
    clgSpy = jest.spyOn(console, 'log')
    jest.clearAllMocks()
  })

  // test('Renders correctly', () => {
  //   expect(wrapper.html()).toMatchSnapshot()
  // })

  test('escribir en el input no debe disparar nada(?)', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
    const input = wrapper.find('input')
    await input.setValue('Hello World')

    expect(clgSpy).toHaveBeenCalledTimes(1)
    expect(getAnswerSpy).not.toHaveBeenCalled()
  })

  test('al escribir el signo de (?) debe disparar el fetch ', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
    const input = wrapper.find('input')
    await input.setValue('Hello World?')
    expect(clgSpy).toHaveBeenCalledTimes(1)
    expect(getAnswerSpy).toHaveBeenCalled()
  })

  test('pruebas en getAnswer', async () => {
    await wrapper.vm.getAnswer()
    const image = wrapper.find('img')

    expect(image.exists()).toBeTruthy()
    expect(wrapper.vm.image).toBe('https://yesno.wtf/assets/yes/2.gif')
    expect(wrapper.vm.answer).toBe('Si')
  })

  test('pruebas en getAnswer - Fallo en el API', async () => {
    fetch.mockImplementationOnce(() => Promise.reject('Api is down'))
    await wrapper.vm.getAnswer()
    const image = wrapper.find('img')

    expect(image.exists()).toBeFalsy()
    expect(wrapper.vm.answer).toBe('No se pudo cargar la API')
  })
})
