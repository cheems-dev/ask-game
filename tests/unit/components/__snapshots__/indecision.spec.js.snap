// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <!--v-if-->
  <div
    class="bg-dark"
  />
  <div
    class="indecision-container"
  >
    <h1>
      Recuerda terminar con un signo de interrogaci√≥n (?)
    </h1>
    <input
      placeholder="Haz una pregunta"
      type="text"
    />
    <!--v-if-->
  </div>
  
</div>
`;
